volumina/tiling/cache.py

```diff
@@ -0,0 +1,295 @@
+import threading
+import warnings
+import logging
+from collections import OrderedDict, defaultdict
+
+import numpy
+from PyQt5.QtWidgets import QGraphicsItem
+
+logger = logging.getLogger()
+
+__all__ = ["TilesCache"]
+
+
+class CachePolicy:
+    def __init__(self, size):
+        self._validate_size(size)
+        self._size = size
+        self._subsribers = []
+
+    @property
+    def size(self):
+        return self._size
+
+    def _validate_size(self, value):
+        if not isinstance(value, int) or value <= 0:
+            raise ValueError("size should non negative integer")
+
+    def set_size(self, value):
+        if self._size == value:
+            return
+
+        self._validate_size(value)
+
+        self._size = value
+
+        for sub in self._subsribers:
+            sub()
+
+    def subscribe(self, fn):
+        self._subsribers.append(fn)
+
+    def unsubsribe(self, fn):
+        try:
+            self._subsribers.remove(fn)
+        except ValueError:
+            pass
+
+
+class MultiCache:
+    """
+    A utility class for caching items in a of a dict-of-dicts
+    """
+
+    def __init__(self, first_uid, default_factory=lambda: None, policy=None):
+        self._policy = policy
+        self._policy.subscribe(self._clean)
+        self._caches = OrderedDict()
+        self._default_factory = default_factory
+        self.add(first_uid)
+
+    def add(self, uid) -> None:
+        if uid not in self._caches:
+            cache = defaultdict(self._default_factory)
+            self._caches[uid] = cache
+        else:
+            raise Exception("MultiCache.add: uid %s is already in use" % str(uid))
```

Comment:
Okay, this has to go, raising exception is just not right! Every code fragment wanting to catch this would have to `except Exception....`

Reply:
I will do follow up pull-request there are some libraries that provide syncornized decorator that do this.